backend/src/main/java/planit/people/preparation/GoogleConnector/GoogleHelper.java

```diff
@@ -40,34 +42,110 @@ public String getRefreshToken() {
-        org.joda.time.DateTime jodaStartDateTime = new org.joda.time.DateTime(startDate);
-        org.joda.time.DateTime jodaStopDateTime = new org.joda.time.DateTime(endDate);
-        Duration durationInMinutes = Duration.standardMinutes(duration);
-        List<Interval> freeBusyForAllAttendee = getFreeBusyForAll(startDate, endDate, refreshTokens);
-        Interval firstInterval = Scheduler.getOneTimeSlotBetweenDatesOfLength(freeBusyForAllAttendee, durationInMinutes, jodaStartDateTime, jodaStopDateTime);
-        Date eventStartDate = firstInterval.getStart().toDate();
-        return new DateTime(eventStartDate);
+    /**
+     * Get event start date. the method will perform the following steps: 
+     * 1 - Get FreeBusy from Google for all registered and invited guests. 
+     * 2 - Using the Free Busy in step 1 get all Free Interval for all invited and registered guests, taken into account the preset availability. 
+     * 3 - Convert the free intervals from step 2 into SchedulingInfo objects. 
+     * 4 - Run the algorithm to find the best time slot from the provided SchedulingInfo records. 
+     * 
+     * @param newEventDetail new event details
+     * @param refreshTokenForAllGuests a map containing the obligation to attend for all registered guests and the refresh token for all of their google accounts. The map structure is the following: Map<Obligation to attend, Map<PlanIt User Id, Set<Refresh Token>>> 
+     * @return DateTime the start datetime of the invite. 
+     * @see GoogleHelper#getFreeBusyIntervalForAll(Date, Date, Map<String, Map<Long, Set<String>>>)
+     * @see GoogleHelper#getFreeIntervalForAll(Map<String, Map<Long, List<Interval>>>, Date, Date, Duration, List<Entity_PresetAvailability>)
+     * @see Converter#convertIntervalMapToListOfSchedulingInfo(Map<String, Map<Long, List<Interval>>>)
+     * @see FreeTimeFinder#getBestStartDate(List<SchedulingInfo>, Duration)
+     */
+    public DateTime getStartDate(DTO_NewEventDetail newEventDetail, Map<String, Map<Long, Set<String>>> refreshTokenForAllGuests) throws ExecutionException, InterruptedException {
+        Duration durationInMinutes = Duration.standardMinutes(newEventDetail.duration());
+        Map<String, Map<Long, List<Interval>>> freeBusyIntervalsForAllUsers = getFreeBusyIntervalForAll(newEventDetail.start_date(), newEventDetail.end_date(), refreshTokenForAllGuests);
+        Map<String, Map<Long, List<Interval>>> freeIntervalsForAllUsers = getFreeIntervalForAll(freeBusyIntervalsForAllUsers, newEventDetail.start_date(), newEventDetail.end_date(), durationInMinutes, newEventDetail.event_preset_detail().preset_availability());
+        List<SchedulingInfo> schedulingInfos = Converter.convertIntervalMapToListOfSchedulingInfo(freeIntervalsForAllUsers);
+        Date startTime = FreeTimeFinder.getBestStartDate(schedulingInfos, durationInMinutes);
+        assert startTime != null;
```

Comment:
[Assertions should never be triggered in production code, and are indicative of a bug or misuse of a code path](https://stackoverflow.com/questions/2758224/what-does-the-java-assert-keyword-do-and-when-should-it-be-used)